src/middlewared/middlewared/plugins/datastore/connection.py

```diff
@@ -27,6 +27,30 @@ class Config:
     engine = None
     connection = None
 
+    @private
+    def handle_constraint_violation(self, row):
+        self.logger.warning("Row %d in table %s violates foreign key constraint on table %s.",
+                            row.rowid, row.table, row.parent)
+
+        if row.table == "directoryservice_idmap_domain" and row.rowid <= 5 and row.parent == "system_certificate":
+            """
+            In commit 5265c8c49f8 a migration was written to use AUTOINCREMENT to ensure id uniqueness.
+            In commit 85f5b97ec9a the aforementioned migration was modified to also fix potential constraint
+            violation in this field.
+
+            Since there was a gap between these two commits, it is impossible that the original
+            migration without the subsequent revision and therefore the user's DB still contains the original
+            constraint violation. This table entry is critical to the proper function of the AD
```

Comment:
I can disagree.

Commit 85f5b97ec9a was added because there was a migration crash described here https://ixsystems.atlassian.net/browse/NAS-111944. Recreating the table with new data was failing because the original table was containing invalid data (specifically `idmap_domain_certificate_id = ''`).

If the user had such an invalid data in their database before it's not possible they had applied `2021-01-20_10-19_autoincrement` successfully before my fix was made. If they did not have such data then both unpatched and patched migration will (or would have) yield the same result.

Reply:
I reproduced the error by installing 21.08-BETA1 on a VM and upgrading to 22.02.4.